### DISCUSSION TITLE
Mocks from snaphots

### DISCUSSION THREAD

**Discussion by AntoineD** [15r10nk/inline-snapshot#97] [Q&A]
Has anyone used snapshots to create mock that replace slow external components?

**Comment by 15r10nk** [comment]
Not Now. You can't read the value of snapshots. 

But that might be something that could be useful.

**Reply by 15r10nk** [reply]
@AntoineD, do you have a specific example in mind?

I'm thinking about how I could design the API and some use cases would help me with that.

**Reply by AntoineD** [reply]
It's still unclear, but let's say we want to mock a slow external function from an external_package. There could be a `mock.patch` like UI that would have 2 behaviors:
- call the external function, record its returned values in a snapshot and return them,
- mock the external function to return the recorded returned values from a snapshot.

The behavior would be chosen from the CLI. A snapshot would depend on the call arguments of the external function.

**Reply by 15r10nk** [reply]
My current Idea looks like this (nothing of this is currently implemented):

``` python
from inline_snapshot import snapshot, createRecorder, Trace


# basic building blocks:
# - createRecorder, to wrap functions or other objects and to record the accesses/calls
# - Trace, to represent the calls which where made.
#   The lambda would be called with some special objects to create the DSL


def external_function(arg):
    return arg + 1


def test_function():
    mocked_external_function, recorder = createRecorder(external_function)

    recorder.save(
        snapshot(
            Trace(
                lambda external_function: [
                    external_function(5) == 6,
                    external_function(2) == 3,
                ]
            )
        )
    )

    mocked_external_function(5)
    mocked_external_function(2)


# the concept could also be extended to classes (and maybe modules)


class external_class:
    def __init__(self, value):
        self.value = value

    def add(self, value):
        return self.value + value

    def sub(self, value):
        return self.value - value


def test_class():
    mocked_class, recorder = createRecorder(external_class)

    recorder.save(
        snapshot(
            Trace(
                lambda external_class: [
                    v1 := external_class(5),
                    v1.add(2) == 7,
                    v1.sub(1) == 4,
                ]
            )
        )
    )

    c = mocked_class(5)
    c.add(2)
    c.sub(1)

```

There could be even 3 modes:
- record: run the external code and create the trace.
- replay: use the trace to provide the results for the mocked functions.
- check: run the external code and check if the trace is correct.

I'm still unsure about some naming and how I want to store multiple recordings in one snapshot/Trace, but I think something like this could be implemented.

The Idea is that a generic design like this could be used to implement different use cases and provide a basic API for other libraries which want to provide some custom snapshot-testing.

Feedback is welcome.

**Reply by AntoineD** [reply]
Thanks a lot @15r10nk!

Would this code be internal to inline-snapshot or written by the test writers?

**Reply by 15r10nk** [reply]
Yes, this code would be written by the test authors. I have updated my example and added the `test_*` functions. 

But I think it could also be useful for libraries to integrate some snapshot tests.
There could be libraries like `inline-snapshot-somelib` that provide functions to test libraries that use `somelib` and record the useful API calls.

But this is all still far in the future :-)

**Reply by AntoineD** [reply]
> Yes, this code would be written by the test authors. I have updated my example and added the `test_*` functions.

In my use case the external function is used by a class but is not exposed and not called directly in the tests.

This is why I had in mind an API similar to [`mock.patch`](https://docs.python.org/3.11/library/unittest.mock.html#unittest.mock.patch) where pretty much only `target` would be passed.

A naive usage example in a test module could be:
```python
from package.module import Class
import pytest
from inline_snapshot import mock

@pytest.fixture
def mock_function():
    # Assuming external_function is imported in package.module global scope
    # and Class uses external_function internally.
    with mock("package.module.external_function"):
        yield

def test_module(mock_function):
    obj = Class(...)
    assert obj.do(...) == ...
```

**Reply by 15r10nk** [reply]
My code above is the basic building block. Other functions can use this and build stuff on top of it.

I think the code bellow should do what you want.

``` python
from mock import patch
from importlib import import_module
from contextlib import contextmanager

@contextmanager
def record(path,trace):

    # lookup the original object
    module_path,attr=path.rsplit(".",1)
    obj=getattr(import_module(module_path),attr)

    # record access to this object
    mocked_object,recorder=createRecorder(obj)
    recorder.save(trace)
    
    # patch the original object
    with patch(path,mocked_object):
        yield


def test_something():
    with record("package.module.external_function",snapshot()):
        obj=Class(...)
        assert obj.do(...) == ...
```

I think that your use case will be very common and it make sense to put something like this into inline-snapshot.

**Reply by AntoineD** [reply]
That looks awesome, how can I help you?

**Reply by 15r10nk** [reply]
I don't know if this is a good starter project. It can be complicated and I'm still thinking about the concrete API. I will not stop you if you want to try it but I don't know if I have much time in the next weeks to help you (I'm moving with my family ...). 

But there are also other less complicated issues which you can solve first if you want, like #58 or #99
